src/net/server/handlers/mul.rs

```diff
@@ -0,0 +1,166 @@
+use crate::helpers::fabric::{ChannelId, MessageChunks, MessageEnvelope};
+use crate::helpers::Identity;
+use crate::net::server::MpcServerError;
+use crate::net::RecordHeaders;
+use crate::protocol::{IPAProtocolStep, QueryId, RecordId, Step};
+use async_trait::async_trait;
+use axum::{
+    body::Bytes,
+    extract::{self, FromRequest, Query, RequestParts},
+    Extension,
+};
+use tokio_util::sync::PollSender;
+
+/// Used in the axum handler to extract the `query_id` and `step` from the path of the request
+pub struct Path<S: Step>(QueryId, S);
+#[async_trait]
+impl<B: Send, S: Step> FromRequest<B> for Path<S> {
+    type Rejection = MpcServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Path((query_id, step)) =
+            extract::Path::<(QueryId, String)>::from_request(req).await?;
+        let step = S::try_from(step).map_err::<serde_json::Error, _>(serde::de::Error::custom)?;
+        Ok(Path(query_id, step))
+    }
+}
+
+/// Injects a permit to send data to the message layer into the Axum request, so that downstream
+/// handlers have simple access to the correct value
+///
+/// For now, stub out the permit logic with just an empty channel
+// pub async fn upstream_middleware_fn<B: Send, S: Step>(
+//     message_stream: MessageStreamExt<S>,
+//     req: Request<B>,
+//     next: Next<B>,
+// ) -> Result<Response, MpcServerError> {
+//     let permit = message_stream.sender.reserve_owned().await?;
+//
+//     let mut req_parts = RequestParts::new(req);
+//     req_parts.extensions_mut().insert(permit);
+//
+//     let req = req_parts.try_into_request()?;
+//
+//     Ok(next.run(req).await)
+// }
+
+/// accepts all the relevant information from the request, and push all of it onto the gateway
+#[allow(clippy::unused_async)] // handler is expected to be async
+#[allow(clippy::cast_possible_truncation)] // length of envelopes array known to be less u32
+pub async fn handler(
+    Extension(mut permit): Extension<PollSender<MessageChunks<IPAProtocolStep>>>,
+    Path(_query_id, step): Path<IPAProtocolStep>,
+    Query(identity): Query<Identity>,
+    RecordHeaders { offset, data_size }: RecordHeaders,
+    body: Bytes,
```

Comment:
we should eventually replace this with `BodyStream` as this is the mechanism in hyper to provide http2 friendly backpressure (window is adjusted after body future is polled).

Reply:
hm...unfortunately this will not be easy to do. `BodyStream` does not implement [`HttpBody`](https://docs.rs/hyper/latest/hyper/body/trait.HttpBody.html) which is required for the `Request<T>`'s T (in Axum). it has a `FromRequest` implementation, but we can't use that because we are already using `Request<T>`